### PR TITLE
[JUJU-344] Wires up the uniter for caas probe support.

### PR DESCRIPTION
**Pull request by tlm**
The uniter now implements probes that are apart of the /startup probe
support in Kubernetes. This PR add this probe into the list of probes
checked.

PR also introduces a new detailed mode for probe endpoints that can
specifically list which probes are passing and failing.

## Checklist

 - [x] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

1. Bootstrap Juju to microk8s
2. Deploy snappass-test
3. Confirm the startup probes are working with:
```
kubectl port-forward snappass-test-0 -n controller-microk8s-localhost 3856:3856
curl http://localhost:3856/startup?detailed=true
```

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: worker/uniter/op_callbacks.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ type operationCallbacks struct {
 func (opc *operationCallbacks) PrepareHook(hi hook.Info) (string, error) {
 	name := string(hi.Kind)
 	switch {
+	case hi.Kind == hooks.Start:
+		opc.u.Probe.SetHasStarted()
```
Comment (hpidcock) [inline worker/uniter/op_callbacks.go:35]: Should be in CommitHook

File: worker/caasprober/probe.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasprober
+
+import (
+	"github.com/juju/juju/observability/probe"
+)
+
+// CaasProbes provides a private internal implementation of CAASProbes.
+type CAASProbes struct {
```
Comment (wallyworld) [inline worker/caasprober/probe.go:11]: if this is a private implementation, it shouldn't be exported
same with the constructor

File: observability/probe/aggregate.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+package probe
+
+import (
+	"github.com/juju/errors"
+)
+
+type Aggregate struct {
+	// Probes is a map of probes to run as part of this aggregate with the key
+	// corresponding to well known name for the probe.
+	Probes map[string]Prober
```
Comment (wallyworld) [inline observability/probe/aggregate.go:13]: if the keys are well known names we should use a type alias

File: observability/probe/aggregate.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package probe
+
+import (
+	"github.com/juju/errors"
+)
+
+type Aggregate struct {
```
Comment (wallyworld) [inline observability/probe/aggregate.go:10]: should this be exported?
we have the NewAggregate function below (which needs a comment)

File: observability/probe/aggregate.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+func NewAggregate() *Aggregate {
+	return &Aggregate{
+		Probes: make(map[string]Prober),
+	}
+}
+
+// Probe implementes Prober Probe
+func (a *Aggregate) Probe() (bool, error) {
+	return a.ProbeWithResultCallback(func(_ string, _ bool, _ error) {})
```
Comment (wallyworld) [inline observability/probe/aggregate.go:24]: if would be helper to define a type for this function or at least name the parameters here to aid readability, or better, define a NoopCallback and pass that in

File: observability/probe/aggregate.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+// in the aggregate calls the provided callback with probe name and result.
+// Useful for building more details reports of what probes are failing and
+// succeeding.
+func (a *Aggregate) ProbeWithResultCallback(
+	cb func(probeKey string, val bool, err error),
+) (bool, error) {
+	rval := true
+	var errVal error
+
+	for k, v := range a.Probes {
```
Comment (wallyworld) [inline observability/probe/aggregate.go:37]: instead of k, v i think name, probe would be clearer

File: observability/probe/aggregate.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		if err != nil && errVal == nil {
+			errVal = err
+			errVal = errors.Annotatef(err, "probe %s", k)
+		} else if err != nil {
+			errVal = errors.Wrap(errVal, errors.Annotatef(err, "probe %s", k))
+		}
+
+		// only change rval if it's currently true. All probes in the aggregate
+		// need to return true to get a true answer.
+		if rval {
```
Comment (wallyworld) [inline observability/probe/aggregate.go:49]: rval = rval && val

?

File: observability/probe/prober.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	"github.com/juju/errors"
+)
+
+// LivenessProber is an interface for probing targets that implement liveness
+// probe support.
+type LivenessProber interface {
+	// LivenessProbe returns the Prober to use for Liveness probes.
+	LivenessProbe() Prober
+}
```
Comment (wallyworld) [inline observability/probe/prober.go:15]: We should keep all of these together, ie readiness and startup are below

I wonder, if we are using a map of well known, named probes, could we just have a single Probe() method, not LivenessProbe(), StartupProbe() etc.
Then the aggregator would just iterate over the supplied, named probers and call Probe() on each one

File: worker/caasprober/controller.go [hunk 9]
```diff
@@ -47,30 +49,30 @@ func (c *Controller) Kill() {
...
-	probes CAASProbes,
+	probes *CAASProbes,
 	mux Mux,
 ) func() error {
 	return func() error {
 		if err := mux.AddHandler(
 			http.MethodGet,
 			k8sconstants.AgentHTTPPathLiveness,
-			ProbeHandler("liveness", probes.LivenessProbe)); err != nil {
+			ProbeHandler("liveness", probes.Liveness)); err != nil {
```
Comment (wallyworld) [inline worker/caasprober/controller.go:59]: we could replace this with (slightly pseudo)
`ProbeHandler("liveness", probes.getProbe("liveness))` (returns a noop prober if not supported)

and use a loop for the AddHandler calls

**Review by wallyworld — COMMENTED** [review]
I've put forward a thought bubble on the approach, we can discuss

File: observability/probe/aggregate.go (hunk 3 above)
Comment (tlm) [reply observability/probe/aggregate.go:13]: They can be anything. Just a canonical name to reference the supplier of the probe. As discussed.

File: observability/probe/aggregate.go (hunk 5 above)
Comment (tlm) [reply observability/probe/aggregate.go:24]: fair enough.

File: observability/probe/aggregate.go (hunk 7 above)
Comment (tlm) [reply observability/probe/aggregate.go:49]: nice pickup

File: observability/probe/prober.go (hunk 8 above)
Comment (tlm) [reply observability/probe/prober.go:15]: Has changed now.

File: worker/caasprober/controller.go [hunk 10]
```diff
@@ -82,10 +84,54 @@ func (c *Controller) makeLoop(
...
+			shouldDetailResponse = val
+		}
+
+		good, err := aggProbe.ProbeWithResultCallback(
+			probe.ProbeResultCallback(func(probeKey string, val bool, err error) {
+				if !shouldDetailResponse {
+					return
+				}
+
+				// We are trying to output 1 line here pre probe called.
```
Comment (hpidcock) [inline worker/caasprober/controller.go:108]: pre or per?

**Review by hpidcock — APPROVED** [review]
Awesome work, thanks.

**Comment by tlm** [comment]
$$merge$$

**Comment by tlm** [comment]
$$merge$$

**Comment by tlm** [comment]
$$merge$$

**Comment by tlm** [comment]
$$merge$$

**Comment by tlm** [comment]
$$merge$$

**Comment by tlm** [comment]
$$merge$$

File: observability/probe/package_test.go [hunk 11]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package probe_test
```
Comment (manadart) [inline observability/probe/package_test.go:4]: We want all these package-local by default.

**Pull request merged**